node-api/backend/validator.go

```diff
@@ -21,80 +21,224 @@
 package backend
 
 import (
+	"slices"
+	"strconv"
+
+	"github.com/berachain/beacon-kit/consensus-types/types"
+	"github.com/berachain/beacon-kit/errors"
 	"github.com/berachain/beacon-kit/node-api/backend/utils"
 	beacontypes "github.com/berachain/beacon-kit/node-api/handlers/beacon/types"
+	"github.com/berachain/beacon-kit/primitives/crypto"
 	"github.com/berachain/beacon-kit/primitives/math"
+	statedb "github.com/berachain/beacon-kit/state-transition/core/state"
 )
 
+type validatorFilters struct {
+	numericIDs []uint64
+	pubkeys    []crypto.BLSPubkey
+}
+
+// parseValidatorIDs parses a slice of string IDs into numeric IDs and pubkeys
+func parseValidatorIDs(ids []string) *validatorFilters {
+	filters := &validatorFilters{
+		numericIDs: make([]uint64, 0, len(ids)),
+		pubkeys:    make([]crypto.BLSPubkey, 0, len(ids)),
+	}
+
+	for _, id := range ids {
+		filters.parseID(id)
+	}
+
+	return filters
+}
+
+// parseID attempts to parse a single ID as either a numeric ID or pubkey
+func (f *validatorFilters) parseID(id string) {
+	// Try parsing as numeric ID first
+	if index, err := strconv.ParseUint(id, 10, 64); err == nil {
+		f.numericIDs = append(f.numericIDs, index)
+		return
+	}
+
+	// Try parsing as pubkey
+	var pubkey crypto.BLSPubkey
+	if err := pubkey.UnmarshalText([]byte(id)); err == nil {
+		f.pubkeys = append(f.pubkeys, pubkey)
+	}
+	// Silently skip invalid IDs
+}
+
+// FilteredValidators will grab all of the validators from the state at the
+// given slot. It will then filter them by the provided ids and statuses.
+func (b Backend) FilteredValidators(
+	slot math.Slot, ids []string, statuses []string,
+) ([]*beacontypes.ValidatorData, error) {
+	st, _, err := b.stateFromSlot(slot)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get state from slot %d", slot)
+	}
+
+	validators, err := st.GetValidators()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get validators")
+	}
+
+	// Parse all IDs and pubkeys once at the start
+	filters := parseValidatorIDs(ids)
+	epoch := b.cs.SlotToEpoch(slot)
+
+	return filterAndBuildValidatorData(st, validators, filters, epoch, statuses)
```

Comment:
q: why we do sanitize ids but not statuses? Shouldn't we check that statuses are among the only ones listed?

Reply:
I believe the echo engine handles this

Reply:
Nice